packages/anoncreds/src/updates/0.3.1-0.4/credentialExchangeRecord.ts

```diff
@@ -36,6 +44,91 @@ export async function migrateCredentialExchangeRecordToV0_4<Agent extends BaseAg
   }
 }
 
+const holderCredentialStates = [
+  CredentialState.Declined,
+  CredentialState.ProposalSent,
+  CredentialState.OfferReceived,
+  CredentialState.RequestSent,
+  CredentialState.CredentialReceived,
+]
+
+export async function getCredentialRole(agent: BaseAgent, credentialRecord: CredentialExchangeRecord) {
+  // Credentials will only have a value when a credential is received, meaning we're the holder
+  if (credentialRecord.credentials.length > 0) {
+    return CredentialRole.Holder
+  }
+  // If credentialRecord.credentials doesn't have any values, and we're also not in state done it means we're the issuer.
+  else if (credentialRecord.state === CredentialState.Done) {
+    return CredentialRole.Issuer
+  }
+  // For these states we know for certain that we're the holder
+  else if (holderCredentialStates.includes(credentialRecord.state)) {
+    return CredentialRole.Holder
+  }
+
+  // Roles only the issuer can have, as sender, when issuing a credential
+  const didCommCredentialIssuerAsSenderRoles = [CredentialState.OfferSent, CredentialState.CredentialIssued]
+
+  // Roles only the holder can have, as sender, when issuing a credential
+  const didCommCredentialHolderAsSenderRoles = [CredentialState.ProposalSent, CredentialState.RequestSent]
+
+  // Roles only the issuer can have, as a recipient, when issuing a credential
+  const didCommCredentialIssuerAsReceiverRoles = [CredentialState.RequestReceived, CredentialState.ProposalReceived]
+
+  // Roles only the holder can have, as a recipient, when issuing a credential
+  const didCommCredentialHolderAsReceiverRoles = [CredentialState.OfferReceived, CredentialState.Done]
+
+  // Fetch the associated didcomm message record
+  const didCommMessageRepository = agent.dependencyManager.resolve(DidCommMessageRepository)
+  const didCommMessageRecord = await didCommMessageRepository.findSingleByQuery(agent.context, {
+    associatedRecordId: credentialRecord.id,
+  })
```

Comment:
There can be multiple didcomm message records retuned for this. Will be good to add a full test that reaches this code with an actual database populated in 0.4 and migrating it to 0.5